docs/annotated_mnist.md

```diff
@@ -1,157 +1,317 @@
-    x = flax.nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
-    x = x.reshape((x.shape[0], -1))
-    x = flax.nn.Dense(x, features=256)
-    x = flax.nn.relu(x)
-    x = flax.nn.Dense(x, features=10)
-    x = flax.nn.log_softmax(x)
-    return x
+
+```python
+def create_optimizer(params, learning_rate, beta):
+  optimizer_def = optim.Momentum(learning_rate=learning_rate, beta=beta)
+  optimizer = optimizer_def.create(params)
+  return optimizer
 ```
 
-`jax.vmap` allows us to define the `cross_entropy_loss`
-function as if it acts on a single sample. `jax.vmap`
-automatically vectorizes code efficiently to run on entire
-batches.
+3. Create a function for parameter initialization:
 
-```py
-@jax.vmap
-def cross_entropy_loss(logits, label):
-  return -logits[label]
+  - Set the initial shape of the kernel (note that JAX and Flax are [row-based](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html#Model-parameters-&-initialization)); and
+  - Initialize the module parameters of your network (`CNN`) with the [`flax.linen.Module.init`](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.init) method using the PRNGKey and parameters (note that the parameters aren't stored with the models).
+
+
+```python
+def get_initial_params(key):
+  init_shape = jnp.ones((1, 28, 28, 1), jnp.float32)
+  initial_params = CNN().init(key, init_shape)['params']
+  return initial_params
 ```
 
-Compute loss and accuracy. We use `jnp` (`jax.numpy`) which can run on
-device (GPU or TPU).
+4. For loss and accuracy metrics, create a separate function:
 
-```py
+
+```python
 def compute_metrics(logits, labels):
-  loss = jnp.mean(cross_entropy_loss(logits, labels))
+  loss = cross_entropy_loss(logits, labels)
   accuracy = jnp.mean(jnp.argmax(logits, -1) == labels)
-  return {'loss': loss, 'accuracy': accuracy}
+  metrics = {
+      'loss': loss,
+      'accuracy': accuracy
+  }
+  return metrics
 ```
 
-`jax.jit` traces the `train_step` function and compiles into fused device
-operations that run on GPU or TPU.
+## The dataset
+
+Define a function that:
+  - Uses TFDS to load and prepare the MNIST dataset; and
+  - Converts the samples to floating-point numbers.
+
+
+```python
+def get_datasets():
+  ds_builder = tfds.builder('mnist')
+  ds_builder.download_and_prepare()
+  # Split into training/test sets
+  train_ds = tfds.as_numpy(ds_builder.as_dataset(split='train', batch_size=-1))
+  test_ds = tfds.as_numpy(ds_builder.as_dataset(split='test', batch_size=-1))
+  # Convert to floating-points
+  train_ds['image'] = jnp.float32(train_ds['image']) / 255.0
+  test_ds['image'] = jnp.float32(test_ds['image']) / 255.0
+  return train_ds, test_ds
+```
+
+## Training and evaluation functions
+
+1. Write a training step function that:
+
+  - Evaluates the neural network given the parameters and a batch of input images with the [`flax.linen.apply`](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.apply) method.
```

Comment:
It's not `flax.linen.apply`. It's `Module.apply`.

Reply:
Thanks for the catch! Much appreciated. (This is why I personally think having direct links to API docs is super useful in tutorials—I couldn't find the  right `apply` given my inexperience with Linen.)

Reply:
Corrected it in 2 places. Thank you again